File: envdiff/analyzers/baseline.py
from __future__ import annotations

import json
from pathlib import Path

from envdiff.models import BaselineEntry, BaselineSnapshot, Finding


def build_baseline_snapshot(findings: tuple[Finding, ...]) -> BaselineSnapshot:
    entries = []
    for finding in findings:
        if not finding.suppression_key:
            continue
        entries.append(
            BaselineEntry(
                suppression_key=finding.suppression_key,
                code=finding.code,
                severity=finding.severity,
                variable_name=finding.variable_name,
                title=finding.title,
                reason=finding.reason,
            )
        )

    ordered = tuple(sorted(entries, key=lambda entry: entry.suppression_key))
    return BaselineSnapshot(entries=ordered)


def write_baseline_snapshot(path: str | Path, findings: tuple[Finding, ...]) -> BaselineSnapshot:
    output_path = Path(path)
    snapshot = build_baseline_snapshot(findings)
    output_path.write_text(
        json.dumps(snapshot.model_dump(mode="json"), indent=2, sort_keys=True) + "\n",
        encoding="utf-8",
    )
    return snapshot


def load_baseline_snapshot(path: str | Path) -> BaselineSnapshot:
    payload = json.loads(Path(path).read_text(encoding="utf-8"))
    return BaselineSnapshot.model_validate(payload)


def load_ignore_keys(path: str | Path) -> set[str]:
    ignore_path = Path(path)
    keys = set()
    for raw_line in ignore_path.read_text(encoding="utf-8").splitlines():
        line = raw_line.strip()
        if not line or line.startswith("#"):
            continue
        keys.add(line)
    return keys


def apply_suppressions(
    findings: tuple[Finding, ...],
    suppression_keys: set[str],
) -> tuple[tuple[Finding, ...], tuple[Finding, ...]]:
    active: list[Finding] = []
    suppressed: list[Finding] = []

    for finding in findings:
        if finding.suppression_key and finding.suppression_key in suppression_keys:
            suppressed.append(finding)
        else:
            active.append(finding)

    return tuple(active), tuple(suppressed)

File: internal/analyzers/baseline.go
package analyzers

import (
	"encoding/json"
	"os"
	"sort"
	"strings"

	"github.com/davisbuilds/envdiff/internal/model"
	"github.com/davisbuilds/envdiff/internal/render"
)

func BuildBaselineSnapshot(findings []model.Finding) model.BaselineSnapshot {
	entries := []model.BaselineEntry{}
	for _, finding := range findings {
		if finding.SuppressionKey == nil {
			continue
		}
		entries = append(entries, model.BaselineEntry{
			Code:           finding.Code,
			Reason:         finding.Reason,
			Severity:       finding.Severity,
			SuppressionKey: *finding.SuppressionKey,
			Title:          finding.Title,
			VariableName:   finding.VariableName,
		})
	}
	sort.SliceStable(entries, func(left int, right int) bool {
		return entries[left].SuppressionKey < entries[right].SuppressionKey
	})
	return model.BaselineSnapshot{Entries: entries, SchemaVersion: model.SchemaVersion}
}

func WriteBaselineSnapshot(path string, findings []model.Finding) (model.BaselineSnapshot, error) {
	snapshot := BuildBaselineSnapshot(findings)
	payload, err := render.MarshalIndent(snapshot)
	if err != nil {
		return model.BaselineSnapshot{}, err
	}
	if err := os.WriteFile(path, append(payload, '\n'), 0o644); err != nil {
		return model.BaselineSnapshot{}, err
	}
	return snapshot, nil
}

func LoadBaselineSnapshot(path string) (model.BaselineSnapshot, error) {
	payload, err := os.ReadFile(path)
	if err != nil {
		return model.BaselineSnapshot{}, err
	}
	var snapshot model.BaselineSnapshot
	if err := json.Unmarshal(payload, &snapshot); err != nil {
		return model.BaselineSnapshot{}, err
	}
	return snapshot, nil
}

func LoadIgnoreKeys(path string) (map[string]struct{}, error) {
	payload, err := os.ReadFile(path)
	if err != nil {
		return nil, err
	}
	keys := map[string]struct{}{}
	for _, rawLine := range strings.Split(string(payload), "\n") {
		line := strings.TrimSpace(rawLine)
		if line == "" || strings.HasPrefix(line, "#") {
			continue
		}
		keys[line] = struct{}{}
	}
	return keys, nil
}

func ApplySuppressions(
	findings []model.Finding,
	suppressionKeys map[string]struct{},
) ([]model.Finding, []model.Finding) {
	active := []model.Finding{}
	suppressed := []model.Finding{}
	for _, finding := range findings {
		if finding.SuppressionKey != nil {
			if _, ok := suppressionKeys[*finding.SuppressionKey]; ok {
				suppressed = append(suppressed, finding)
				continue
			}
		}
		active = append(active, finding)
	}
	return active, suppressed
}
